python/cudf/cudf/core/udf/__init__.py

```diff
@@ -47,7 +47,14 @@
         utils.JIT_SUPPORTED_TYPES |= STRING_TYPES
         _supported_masked_types |= {string_view, udf_string}
 
-        utils.launch_arg_getters[cudf_str_dtype] = column_to_string_view_array
+        def column_to_string_view_array_init_heap(col):
+            # lazily allocate heap only when a string needs to be returned
```

Comment:
The structure of this code makes it look like idempotency is its main goal, not laziness.

`strings_udf.set_malloc_heap_size()` should only ever be called once, I think?

Maybe this function or module in cudf (and not strings_udf) should own the “is it allocated already?” logic via a cache or global of some kind, and then the strings_udf logic doesn’t need to read env vars for the allocation size, etc.

Reply:
This is not an allocation. There is currently a malloc-heap-size by default (set to about 8MB on the GPUs I've tested on). This updates the heap size to a new value so "already allocated" is not the goal.

Reply:
Is `set_malloc_heap_size` intended to be a public API that can be called by a user, or is it an internal function? It's not entirely clear what the public API is here because this file doesn't define an `__all__`. If the function is just for internal use, why does it need to be called every time someone accesses a string column rather than just calling `cudaDeviceSetLimit` once when the package is imported (or after we can guarantee that the CUDA context/etc is initialized if that is the requirement that we're currently addressing by doing this "lazily")?

Reply:
I think what we want to happen is:
- The first time a user runs a string_udf, set the heap size to either default or what is retrieved from their environment. 
- The next time a user runs a string_udf, do not set the heap size.
- Reset the heap the next time the user runs a string_udf iff the user sets the value of `STRING_UDF_HEAP_SIZE` to something else during the session.
Does this seems reasonable?

Reply:
Seems we have 2 comment chains asking about the same thing.
https://github.com/rapidsai/cudf/pull/12094#discussion_r1024471017

Reply:
> - The first time a user runs a string_udf, set the heap size to either default or what is retrieved from their environment. 

Yes.

> - The next time a user runs a string_udf, do not set the heap size. Or at least early return.

The "set" function should not be called. Avoid designs that try to set but exit early.

> - Reset the heap the next time the user runs a string_udf iff the user sets the value of `STRING_UDF_HEAP_SIZE` to something else during the session.

No. The `STRING_UDF_HEAP_SIZE` should not be modifiable (I argued separately that that variable should not exist at all). If the user needs to be able to set a specific size, then the "set" function should be public and the user should be able to call it explicitly with immediate effect (not waiting until the next UDF execution, because memory demands can change in the meantime).

Reply:
With the changes in https://github.com/rapidsai/cudf/commit/af7cd9f85b2afa0a6b92f53e4abffbb538c9d02f I still have `heap_size` available as a module level attribute of `strings_udf`, what do you think about checking if `heap_size == 0` within the getter to determine if the function need be called or not?

Reply:
I would use a cached function for this. Following ideas from https://mail.python.org/archives/list/python-ideas@python.org/thread/5OR3LJO7LOL6SC4OOGKFIVNNH4KADBPG/

```python
from functools import lru_cache

# Only executes the body of the function on the first call
once = lru_cache(maxsize=None)

@once
def set_initial_malloc_heap_size():
    strings_udf.set_malloc_heap_size()

def column_to_string_view_array_init_heap(col):
    set_initial_malloc_heap_size()
    return column_to_string_view_array(col)
```

Reply:
thanks this worked perfectly!